gms/gossiper.cc

```diff
@@ -571,12 +571,32 @@ future<> gossiper::send_gossip(gossip_digest_syn message, std::set<inet_address>
 future<> gossiper::do_apply_state_locally(gms::inet_address node, endpoint_state remote_state, bool listener_notification) {
     // If state does not exist just add it. If it does then add it if the remote generation is greater.
     // If there is a generation tie, attempt to break it by heartbeat version.
+    locator::host_id remote_host_id = remote_state.get_host_id();
+    locator::host_id local_host_id;
+    const auto& addr = node;
     auto permit = co_await this->lock_endpoint(node, null_permit_id);
     auto es = this->get_endpoint_state_ptr(node);
     if (es) {
+        local_host_id = es->get_host_id();
         endpoint_state local_state = *es;
         auto local_generation = local_state.get_heart_beat_state().get_generation();
         auto remote_generation = remote_state.get_heart_beat_state().get_generation();
+
+        if (!local_host_id) {
+            if (!remote_host_id) {
+                logger.warn("Node {} sent null host_id in remote state", addr);
+            } else {
+                logger.info("Node {} host_id is {}", addr, remote_host_id);
+            }
+        } else {
+            if (remote_host_id) {
+                if (local_host_id != remote_host_id) {
+                    logger.warn("Node {} changed host_id from {} to {}", addr, local_host_id, remote_host_id);
+                    co_return co_await this->handle_major_state_change(node, std::move(remote_state), permit.id());
```

Comment:
When searching using the endpoint this would be the "replace with same ip address" case (and the log message sucks, need to improve it.). Once we search by host_id first we should find no endpoint_state for the new node, so `local_host_id != remote_host_id` should never happen and would indicate a bug.

Reply:
Yes, so by the end of the series, this should be impossible. If we don't allow null host ids, this will be explicit in the code and there will be no need for code like this.